sdccc/src/main/java/com/draeger/medical/sdccc/manipulation/precondition/impl/ConditionalPreconditions.java

```diff
@@ -349,6 +352,189 @@ static boolean manipulation(final Injector injector) throws PreconditionExceptio
+                LOG.error("No removable MdsDescriptors could be found via the GetRemovableDescriptorsOfType "
+                        + "manipulation. Please check if the test case applying this precondition is applicable to "
+                        + "your device and if the GetRemovableDescriptorsOfType manipulation has been implemented "
+                        + "correctly.");
+            }
+
+            final List<String> absentRemovableMdsDescriptors = new ArrayList<>();
+            final List<String> presentRemovableMdsDescriptors = new ArrayList<>();
+            for (String removableMdsDescriptor : removableMdsDescriptors) {
+                final Optional<AbstractDescriptor> descOpt = mdibAccess.getDescriptor(removableMdsDescriptor);
+                if (descOpt.isEmpty()) {
+                    absentRemovableMdsDescriptors.add(removableMdsDescriptor);
+                } else {
+                    presentRemovableMdsDescriptors.add(removableMdsDescriptor);
+                }
+            }
+
+            // absent:        present:
+            // 1. insert
+            // 2. update      1. update
+            // 3. remove      2. remove
+            //                3. insert
+
+            // priorities:
+            // 1. try all present descriptors
+            // 2. try all absent descriptors
+            // 3. when none worked -> ERROR
+
+            // 1. try all present descriptors
+            boolean success = false;
+            for (String initiallyPresentMdsDescriptor : presentRemovableMdsDescriptors) {
+                if (triggerAllDescriptorUpdatesForInitiallyPresentMdsDescriptor(
+                        initiallyPresentMdsDescriptor, manipulations)) {
+                    success = true;
+                    break;
+                }
+            }
+            if (!success) {
+                for (String initiallyAbsentMdsDescriptor : absentRemovableMdsDescriptors) {
+                    if (triggerAllDescriptorUpdatesForInitiallyAbsentMdsDescriptor(
+                            initiallyAbsentMdsDescriptor, manipulations)) {
+                        success = true;
+                        break;
+                    }
+                }
+            }
+            if (!success) {
+                // all options exhausted
+                LOG.error("Unable to find any MdsDescriptors using the GetRemovableDescriptorsOfType() manipulation "
+                        + "that can be inserted, updated and removed."
+                        + "Please check if the test case applying this precondition is applicable to your device and if the "
+                        + "GetRemovableDescriptorsOfType, InsertDescriptor, RemoveDescriptor, and TriggerDescriptorUpdate "
+                        + "manipulations have been implemented correctly.");
+                return false;
+            }
+
+            return true;
+        }
+
+        private static boolean triggerAllDescriptorUpdatesForInitiallyAbsentMdsDescriptor(
```

Comment:
Isn't this method and _triggerAllDescriptorUpdatesForInitiallyPresentMdsDescriptor_ too strict? It would also be okay to support only a subset of the manipulations

Reply:
After some discussion we decided to modify the Precondition design in order to
a) avoid unnecessary Manipulation calls, and
b) make the test less strict: Devices that do not have a particular removable MdsDescriptor supporting all three Manipulations (insert, update, delete) can still pass the test as long as each of those is supported by at least one of their removable MdsDescriptors.